backend-api/app/routes/subscriptions.py

```diff
@@ -1,63 +1,84 @@
+from ..schemas.subscriptions import (
+    SubscriptionCreate,
+    SubscriptionRead,
+)
+from ..utils.date import date_to_epoch
+from ..blockchain.subscription import consumer_purchase_tokens
+from ..blockchain.permissions import consumer_subscriptions
 
 router = APIRouter()
 
-subscription_dict = {
-    "creation_date": "2023-07-20",
-    "expiration_date": "2023-07-20",
-    "active": True,
-}
-
 
 @router.post("/", status_code=status.HTTP_201_CREATED)
-async def create_subscriptions(subscriptionIds: list[int]):
-    return [
-        SubscriptionRead(id=random.randint(1, 1000), **subscription_dict)
-        for _ in range(len(subscriptionIds))
-    ]
-
-
-@router.get("/")
-async def read_subscriptions():
-    return [
-        SubscriptionRead(id=random.randint(1, 1000), **subscription_dict)
-        for _ in range(5)
-    ]
-
-
-@router.patch("/")
-async def update_subscriptions(subscriptions: list[SubscriptionUpdate]):
-    return [
-        SubscriptionRead(
-            id=random.randint(1, 1000),
-            **subscription_dict,
+async def create_subscriptions(subscriptions: SubscriptionCreate, request: Request):
+    # Check for a valid expiration date
+    current_date = datetime.now().date()
+    expiration_date = subscriptions.expiration_date
+    if expiration_date <= current_date:
+        raise HTTPException(
+            status_code=status.HTTP_400_BAD_REQUEST, detail="Invalid expiration date"
         )
-        for _ in range(len(subscriptions))
-    ]
+
+    # Perform the token purchases through the smart contract
+    consumer_purchase_tokens(
+        request.app.state.token_contract,
+        subscriptions.consumer_eth_address,
+        subscriptions.producer_eth_addresses,
+        date_to_epoch(current_date),
+        date_to_epoch(expiration_date),
+    )
+
+    subscriptions = consumer_subscriptions(
+        request.app.state.token_contract, subscriptions.consumer_eth_address
+    )
+
+    return SubscriptionRead(subscriptions=subscriptions)
+
+
+@router.get("/{consumer_eth_address}", status_code=status.HTTP_200_OK)
+async def read_subscriptions(consumer_eth_address: str, request: Request):
+    subscriptions = consumer_subscriptions(
+        request.app.state.token_contract, consumer_eth_address
+    )
+
+    return SubscriptionRead(subscriptions=subscriptions)
+
+
+@router.patch("/", status_code=status.HTTP_204_NO_CONTENT)
+async def update_subscriptions(consumer_eth_address: str):
+    # TODO: Implement later
+
+    return None
+
+
+@router.delete("/{consumer_eth_address}", status_code=status.HTTP_204_NO_CONTENT)
+async def delete_subscriptions(consumer_eth_address: str):
+    # TODO: Implement later
+
+    return None
 
 
-@router.get("/available")
-async def read_subscriptions_available():
-    return [str(random.randint(1, 1000)) for _ in range(5)]
+# @router.get("/available")
+# async def read_subscriptions_available():
+#     return [str(random.randint(1, 1000)) for _ in range(5)]
```

Comment:
How did you want the browser extension to find the subscriptions available to be purchased.

Reply:
Ah sorry I didn't realize what this endpoint was for, I'll uncomment it!